gateway/webhook/integration_test.go

```diff
@@ -0,0 +1,361 @@
+
+	"github.com/rudderlabs/rudder-go-kit/stats/memstats"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/gateway"
+)
+
+type testSetup struct {
+	context testContext
+	cases   []testCase
+}
+
+type testContext struct {
+	Now       time.Time
+	RequestIP string `json:"request_ip"`
+}
+
+type testCase struct {
+	Name        string
+	Description string
+	Input       testCaseInput
+	Output      testCaseOutput
+
+	config backendconfig.SourceT
+}
+
+type testCaseInput struct {
+	Request testCaseRequest
+}
+type testCaseRequest struct {
+	Method   string
+	RawQuery string `json:"query"`
+	Headers  map[string]string
+	Body     json.RawMessage
+}
+
+type testCaseOutput struct {
+	Response testCaseResponse
+	Queue    []json.RawMessage
+	ErrQueue []json.RawMessage `json:"err_queue"`
+}
+
+type testCaseResponse struct {
+	Body       json.RawMessage
+	StatusCode int `json:"status"`
+}
+
+func TestIntegrationWebhook(t *testing.T) {
+	ctx, _ := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGINT)
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Minute)
+	defer cancel()
+
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	var (
+		p                    *postgres.Resource
+		transformerContainer *transformertest.Resource
+	)
+
+	g, _ := errgroup.WithContext(ctx)
+	g.Go(func() (err error) {
+		p, err = postgres.Setup(pool, t)
+		if err != nil {
+			return fmt.Errorf("starting postgres: %w", err)
+		}
+		return nil
+	})
+	g.Go(func() (err error) {
+		transformerContainer, err = transformertest.Setup(pool, t)
+		if err != nil {
+			return fmt.Errorf("starting transformer: %w", err)
+		}
+		return nil
+	})
+	err = g.Wait()
+	require.NoError(t, err)
+
+	var gw gateway.Handle
+
+	conf := config.New()
+	logger := logger.NOP
+	stat, err := memstats.New()
+	require.NoError(t, err)
+
+	conf.Set("Gateway.enableSuppressUserFeature", false)
+
+	gatewayDB := jobsdb.NewForReadWrite(
+		"gateway",
+		jobsdb.WithDBHandle(p.DB),
+	)
+	require.NoError(t, gatewayDB.Start())
+	defer gatewayDB.TearDown()
+
+	errDB := jobsdb.NewForReadWrite(
+		"err",
+		jobsdb.WithDBHandle(p.DB),
+	)
+	require.NoError(t, errDB.Start())
+	defer errDB.TearDown()
+
+	var (
+		rateLimiter        throttler.Throttler
+		versionHandler     func(w http.ResponseWriter, r *http.Request)
+		streamMsgValidator func(message *stream.Message) error
+		application        app.App
+	)
+
+	transformerFeaturesService := transformer.NewFeaturesService(ctx, conf, transformer.FeaturesServiceOptions{
+		PollInterval:             config.GetDuration("Transformer.pollInterval", 10, time.Second),
+		TransformerURL:           transformerContainer.TransformerURL,
+		FeaturesRetryMaxAttempts: 10,
+	})
+	t.Setenv("DEST_TRANSFORM_URL", transformerContainer.TransformerURL)
+
+	<-transformerFeaturesService.Wait()
+
+	bcs := make(map[string]backendconfig.ConfigT)
+
+	testSetup := loadTestSetup(t)
+
+	for i, tc := range testSetup.cases {
```

Comment:
[optional] I found this a bit weird that you're iterating over the test cases twice. Maybe you can create a function for this an call it at the beginning of each iteration in the other loop?

Reply:
We need to prepare backendconfig before initialising the service